prometheus_juju_backup_all_exporter/collector.py

```diff
@@ -1,169 +1,116 @@
-            new_metrics.append(new_metric)
-        self._metrics = tuple(new_metrics)
-
-    def collect(self) -> Iterator[Metric]:
-        """Fetch data and update the internal metrics.
-
-        This is a callback method that is used internally within
-        `prometheus_client` every time the exporter server is queried. There is
-        not return values for this method but it needs to yield all the metrics.
-
-        Yields:
-            metrics: the internal metrics.
-        """
-        self._update_metrics()
-        for metric in self._metrics:
-            yield metric
-
-
-class BackupStateCollector(CollectorBase):  # pylint: disable=R0903
-    """Collector for backup state."""
-
-    def __init__(self, config: Config, *metrics: Metric) -> None:
-        """Initialize the class."""
-        super().__init__(*metrics)
-        self.config = config
+            Specification(
+                name="juju_backup_all_backup_purged_total",
+                documentation="The number of purged backups.",
+                labels=[],
+                metric_class=CounterMetricFamily,
+            ),
+            Specification(
+                name="juju_backup_all_backup_completed_total",
+                documentation="The number of completed backups.",
+                labels=[],
+                metric_class=CounterMetricFamily,
+            ),
+        ]
+
+    def fetch(self) -> List[Payload]:
+        """Load the backup event data."""
+        backup_event = BackupEvent(self.config)
+        return [
+            Payload(
+                name="juju_backup_all_backup_failed_total",
+                labels=[],
+                value=float(backup_event.failed),
+            ),
+            Payload(
+                name="juju_backup_all_backup_purged_total",
+                labels=[],
+                value=float(backup_event.purged),
+            ),
+            Payload(
+                name="juju_backup_all_backup_completed_total",
+                labels=[],
+                value=float(backup_event.completed),
+            ),
+        ]
 
-    def _fetch(self) -> Dict[str, List[Dict[str, Any]]]:
-        """User defined _fetch method used by `self._update_metrics()`."""
-        backup_state = BackupState(self.config)
-        return {
-            "failed_metric": [
-                {"labels": [], "value": float(backup_state.failed)},
-            ],
-            "purged_metric": [
-                {"labels": [], "value": float(backup_state.purged)},
-            ],
-            "completed_metric": [
-                {"labels": [], "value": float(backup_state.completed)},
-            ],
-        }
+    def process(self, payloads: List[Payload], datastore: Dict[str, Payload]) -> List[Payload]:
+        """Process the backup event data."""
+        # Increments the counter according to the payload.
+        for payload in payloads:
+            payload.value += datastore[payload.uuid].value
```

Comment:
Can you provide example what kind of number could be see in `.value`? Counters are usually increased by 1, so I'm not sure what is happening here. Thanks

Reply:
Counters can be increased by any (positive) value, and it's not necessary 1. it's also supported in [prometheus python client](https://github.com/prometheus/client_python#counter) In here, for charm j-b-a, the return payload (purged, completed, failed) are all number >= 0.

Reply:
Thanks for clarification, I spotted it in code and without context it's looks suspicious to me.